sys/cddl/compat/opensolaris/sys/vnode.h

```diff
@@ -268,7 +268,11 @@ vn_rename(char *from, char *to, enum uio_seg seg)
 
 	ASSERT(seg == UIO_SYSSPACE);
 
-	return (kern_renameat(curthread, AT_FDCWD, from, AT_FDCWD, to, seg));
+    return (kern_renameat(curthread, AT_FDCWD,
+    		(__cheri_tocap const char * __capability) from,
```

Comment:
Does this produce an invalid capability?

Reply:
yes, vn_rename needs to be taking capabilities. (as does vn_remove below)

Reply:
In this case, should I create a capability with `__USER_CAP` then?

Reply:
No, you would need to change the vn_*() helpers to take __capability arguments and construct capabilities at the call site if required.

That being said, I fear you've wondered into the weeds as looks to be ZFS-specific code.

Reply:
does the annotation `__cheri_tocap` create an unbounded capability?
btw, this should have been fixed [here](https://github.com/CTSRD-CHERI/cheribsd/pull/415).

Reply:
It creates an unbounded capability (IIRC unless the size is visible to the compiler).  That's fine in practice when your only alternative is to call strlen because either way you're relying on NUL termination.

Reply:
Well, it’s not quite the same. By using strlen here to bound the cap you ensure that nobody else can silently screw up ignore the NUL terminator. I view it as still worthwhile in general.